src/sfeAS7331.cpp

```diff
@@ -0,0 +1,1107 @@
+#include "sfeAS7331.h"
+
+bool SfeAS7331Driver::begin(const uint8_t &deviceAddress, sfeTkIBus *theBus)
+{
+    // Nullptr check.
+    if (!_theBus && !theBus)
+        return false;
+
+    // Set the internal bus pointer, overriding current bus if it exists.
+    if (theBus != nullptr)
+        setCommunicationBus(theBus);
+
+    // If the address passed in isn't the default, set the new address.
+    if (kDefaultAS7331Addr != deviceAddress)
+        setDeviceAddress(deviceAddress);
+
+    // Get the device setup and ready.
+    return runDefaultSetup();
+}
+
+uint8_t SfeAS7331Driver::getDeviceID(void)
+{
+    // Nullptr check.
+    if (!_theBus)
+        return 0;
+
+    // Check sensor's operating mode.
+    sfe_as7331_reg_cfg_osr_t osr;
+    if (kSTkErrOk != getOSR(osr))
+        return 0;
+
+    bool needsToBeChangedBack = false;
+
+    // Change if necessary.
+    if (osr.dos != DEVICE_MODE_CFG)
+    {
+        needsToBeChangedBack = true;
+
+        osr.dos = DEVICE_MODE_CFG;
+
+        if(kSTkErrOk != setOSR(osr))
+            return 0;
+    }
+    
+    uint8_t devID;
+
+    // Read the device ID register, if it errors then return 0.
+    if (kSTkErrOk != _theBus->readRegisterByte(kSfeAS7331RegCfgAgen, devID))
+        return 0;
+
+    // If we changed it at first, change it back.
+    if(needsToBeChangedBack)
+    {
+        if(kSTkErrOk != setOperationMode(DEVICE_MODE_MEAS))
+            return 0;
+    }
+
+    return devID;
+}
+
+void SfeAS7331Driver::setCommunicationBus(sfeTkIBus *theBus)
+{
+    _theBus = theBus;
+}
+
+void SfeAS7331Driver::setDeviceAddress(const uint8_t &deviceAddress)
+{
+    switch(deviceAddress)
+    {
+        // If it's any of the allowed addresses, set it.
+        case kDefaultAS7331Addr:
+        case kSecondaryAS7331Addr:
+        case kTertiaryAS7331Addr:
+        case kQuaternaryAS7331Addr:
+            _devAddress = deviceAddress;
+            break;
+        default: // Otherwise set it to the default. No disallowed addresses.
+            _devAddress = kDefaultAS7331Addr;
```

Comment:
I think the default case should just do nothing, and leave the address at whatever it was.